File: studyq-backend/services/scoring_service.py
from models.question import Question


def calculate_score(quiz_id: int, answers: dict) -> dict:
    """
    Compare submitted answers to correct answers.

    Parameters
    ----------
    quiz_id : int
    answers : dict  {question_id (str/int): chosen_option (str 'A'|'B'|'C'|'D')}

    Returns
    -------
    dict with score, total, accuracy, and per-question feedback
    """
    questions = Question.query.filter_by(quiz_id=quiz_id).all()
    total = len(questions)
    score = 0
    feedback = []

    for q in questions:
        submitted = answers.get(str(q.id), "").strip().upper()
        correct = q.correct_answer.upper()
        is_correct = submitted == correct
        if is_correct:
            score += 1
        feedback.append({
            "question_id": q.id,
            "question_text": q.question_text,
            "submitted": submitted or None,
            "correct": correct,
            "is_correct": is_correct,
        })

    accuracy = round((score / total) * 100, 1) if total > 0 else 0.0
    return {
        "score": score,
        "total_questions": total,
        "accuracy": accuracy,
        "feedback": feedback,
    }

File: studyq-backend/services/scoringService.js
const Question = require('../models/Question');

const calculateScore = async (quizId, answers) => {
  const questions = await Question.find({ quiz_id: quizId });
  const total = questions.length;
  let score = 0;
  const feedback = [];

  for (const q of questions) {
    const qIdStr = q._id.toString();
    const submitted = (answers[qIdStr] || '').trim().toUpperCase();
    const correct = q.correct_answer.toUpperCase();
    const isCorrect = submitted === correct;
    
    if (isCorrect) {
      score += 1;
    }
    
    feedback.push({
      question_id: qIdStr,
      question_text: q.question_text,
      submitted: submitted || null,
      correct: correct,
      is_correct: isCorrect
    });
  }

  const accuracy = total > 0 ? Number(((score / total) * 100).toFixed(1)) : 0.0;
  return {
    score,
    total_questions: total,
    accuracy,
    feedback
  };
};

module.exports = {
  calculateScore
};
